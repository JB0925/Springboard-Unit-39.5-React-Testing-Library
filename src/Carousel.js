import React, { useState } from "react";
import "./Carousel.css";
import image1 from "./image1.jpg";
import image2 from "./image2.jpg";
import image3 from "./image3.jpg";
import Card from "./Card";

function Carousel(props) {
  const [cardIdx, setCardIdx] = useState(0);
  const card = props.cardData[cardIdx];
  const total = props.cardData.length;

  // I rewrote this function to satisfy requirement 3 and 4 of Unit 39.5
  // though it is a little bit different, I believe it handles the situation 
  // more elegantly and the test I wrote for it passes.
  const changePictureSrc = (evt) => {
    const deckLength = props.cardData.length;

    if (evt.target.id === "left-arrow") {
      cardIdx === 0 ? setCardIdx(deckLength - 1) : setCardIdx(cardIdx - 1);
      return;
    };

    cardIdx === deckLength - 1 ? setCardIdx(0) : setCardIdx(cardIdx + 1);
  };

  return (
    <div className="Carousel">
      <h1>{props.title}</h1>
      <div className="Carousel-main">
        <i
          className="fas fa-chevron-circle-left fa-2x"
          onClick={changePictureSrc}
          data-testid="left-arrow"
          id="left-arrow"
        />
        <Card
          caption={card.caption}
          src={card.src}
          currNum={cardIdx + 1}
          totalNum={total}
        />
        <i
          className="fas fa-chevron-circle-right fa-2x"
          onClick={changePictureSrc}
          data-testid="right-arrow"
          id="right-arrow"
        />
      </div>
    </div>
  );
}

Carousel.defaultProps = {
  cardData: [
    {
      src: image1,
      caption: "Photo by Richard Pasquarella on Unsplash"
    },
    {
      src: image2,
      caption: "Photo by Pratik Patel on Unsplash"
    },
    {
      src: image3,
      caption: "Photo by Josh Post on Unsplash"
    }
  ],
  title: "Shells from far away beaches."
};

export default Carousel;
