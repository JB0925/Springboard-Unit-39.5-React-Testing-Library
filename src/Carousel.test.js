import React from "react";
import { render, fireEvent } from "@testing-library/react";
import Carousel from "./Carousel";
import Card from "./Card";
import App from "./App";
import { expect } from "@jest/globals";

it("renders the Card component", () => {
  render(<Card 
           caption="hi hello there"
           src="https://www.hello.com"
           currNum={5}
           totalNum={15}/>);
});

it("renders the Carousel component", () => {
  render(<Carousel />);
});

it("matches the snapshot", () => {
  const { asFragment } = render(<App/>);
  expect(asFragment()).toMatchSnapshot();
});

it("matches the Card snapshot", () => {
  const { asFragment } = render(<Card />);
  expect(asFragment()).toMatchSnapshot();
});

it("matches the Carousel snapshot", () => {
  const { asFragment } = render(<Carousel />);
  expect(asFragment()).toMatchSnapshot();
});

it("works when you click on the right arrow", function() {
  const { queryByTestId, queryByAltText } = render(<Carousel />);

  // expect the first image to show, but not the second
  expect(queryByAltText("Photo by Richard Pasquarella on Unsplash")).toBeInTheDocument();
  expect(queryByAltText("Photo by Pratik Patel on Unsplash")).not.toBeInTheDocument();

  // move forward in the carousel
  const rightArrow = queryByTestId("right-arrow");
  fireEvent.click(rightArrow);

  // expect the second image to show, but not the first
  expect(queryByAltText("Photo by Richard Pasquarella on Unsplash")).not.toBeInTheDocument();
  expect(queryByAltText("Photo by Pratik Patel on Unsplash")).toBeInTheDocument();
  
  // click past the last image
  fireEvent.click(rightArrow);
  fireEvent.click(rightArrow);

  // expect the first image to show, but not the last
  expect(queryByAltText("Photo by Richard Pasquarella on Unsplash")).toBeInTheDocument();
  expect(queryByAltText("Photo by Josh Post on Unsplash")).not.toBeInTheDocument();
});

it("works when you click on the left arrow", function() {
  const { queryByTestId, queryByAltText } = render(<Carousel />);

  // expect the first image to show, but not the second
  expect(queryByAltText("Photo by Richard Pasquarella on Unsplash")).toBeInTheDocument();
  expect(queryByAltText("Photo by Pratik Patel on Unsplash")).not.toBeInTheDocument();

  // move forward in the carousel
  const leftArrow = queryByTestId("left-arrow");
  fireEvent.click(leftArrow);

  // expect the last image to show, but not the first
  expect(queryByAltText("Photo by Richard Pasquarella on Unsplash")).not.toBeInTheDocument();
  expect(queryByAltText("Photo by Josh Post on Unsplash")).toBeInTheDocument();
});
