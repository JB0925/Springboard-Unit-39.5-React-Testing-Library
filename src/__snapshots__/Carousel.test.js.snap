// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches the Card snapshot 1`] = `
<DocumentFragment>
  <div
    class="Card"
  >
    <h4
      class="Card-title"
    />
    <img
      class="Card-image"
    />
    <small
      class="Card-small"
    >
      Image  of .
    </small>
  </div>
</DocumentFragment>
`;

exports[`matches the Carousel snapshot 1`] = `
<DocumentFragment>
  <div
    class="Carousel"
  >
    <h1>
      Shells from far away beaches.
    </h1>
    <div
      class="Carousel-main"
    >
      <i
        class="fas fa-chevron-circle-left fa-2x"
        data-testid="left-arrow"
        id="left-arrow"
      />
      <div
        class="Card"
      >
        <h4
          class="Card-title"
        >
          Photo by Richard Pasquarella on Unsplash
        </h4>
        <img
          alt="Photo by Richard Pasquarella on Unsplash"
          class="Card-image"
          src="image1.jpg"
        />
        <small
          class="Card-small"
        >
          Image 1 of 3.
        </small>
      </div>
      <i
        class="fas fa-chevron-circle-right fa-2x"
        data-testid="right-arrow"
        id="right-arrow"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <div
      class="Carousel"
    >
      <h1>
        Shells from far away beaches.
      </h1>
      <div
        class="Carousel-main"
      >
        <i
          class="fas fa-chevron-circle-left fa-2x"
          data-testid="left-arrow"
          id="left-arrow"
        />
        <div
          class="Card"
        >
          <h4
            class="Card-title"
          >
            Photo by Richard Pasquarella on Unsplash
          </h4>
          <img
            alt="Photo by Richard Pasquarella on Unsplash"
            class="Card-image"
            src="image1.jpg"
          />
          <small
            class="Card-small"
          >
            Image 1 of 3.
          </small>
        </div>
        <i
          class="fas fa-chevron-circle-right fa-2x"
          data-testid="right-arrow"
          id="right-arrow"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
